fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,1066 @@
+	t.C.adjacentTriangles = t.C.adjacentTriangles.remove(t)
+	t.A.adjacentTriangles = t.A.adjacentTriangles.appendT(t1, t3)
+	t.B.adjacentTriangles = t.B.adjacentTriangles.appendT(t1, t2)
+	t.C.adjacentTriangles = t.C.adjacentTriangles.appendT(t2, t3)
+	if d.hierarchy {
+		t.children = append(t.children, t1, t2, t3)
+	}
+	d.triangles = append(d.triangles, t1, t2, t3)
+	// change the edges so it is a valid delaunay triangulation
+	d.validateEdge(t1, p)
+	d.validateEdge(t2, p)
+	d.validateEdge(t3, p)
+}
+
+// insertPonE inserts a point on an edge
+func (d *Delaunay) insertPonE(t1 *Triangle, p *Point) {
+	// find second triangle adjacent to edge
+	var t2 *Triangle
+	for _, t2 = range d.triangles {
+		_, edge := p.inTriangle(t2)
+		if edge && !t2.Equals(t1) {
+			break
+		}
+	}
+	// find points opposite and adjacent to the edge
+	var p1, p2, pO1, pO2 *Point
+	switch {
+	case !t1.A.Equals(t2.A) && !t1.A.Equals(t2.B) && !t1.A.Equals(t2.C):
+		p1 = t1.A
+		pO1 = t1.B
+		pO2 = t1.C
+	case !t1.B.Equals(t2.A) && !t1.B.Equals(t2.B) && !t1.B.Equals(t2.C):
+		p1 = t1.B
+		pO1 = t1.A
+		pO2 = t1.C
+	case !t1.C.Equals(t2.A) && !t1.C.Equals(t2.B) && !t1.C.Equals(t2.C):
+		p1 = t1.C
+		pO1 = t1.B
+		pO2 = t1.A
+	default:
+		panic(fmt.Errorf("delaunay: triangle T1%v doesn't have points not in T2%v", t1, t2))
+	}
+	switch {
+	case !t2.A.Equals(t1.A) && !t2.A.Equals(t1.B) && !t2.A.Equals(t1.C):
+		p2 = t2.A
+	case !t2.B.Equals(t1.A) && !t2.B.Equals(t1.B) && !t2.B.Equals(t1.C):
+		p2 = t2.B
+	case !t2.C.Equals(t1.A) && !t2.C.Equals(t1.B) && !t2.C.Equals(t1.C):
+		p2 = t2.C
+	default:
+		panic(fmt.Errorf("delaunay: triangle T2%v doesn't have points not in T1%v", t2, t1))
+	}
+	// form four new triangles
+	nt1 := NewTriangle(p1, p, pO1)
+	nt1.inD = true
+	nt2 := NewTriangle(p, p2, pO1)
+	nt2.inD = true
+	nt3 := NewTriangle(p1, p, pO2)
+	nt3.inD = true
+	nt4 := NewTriangle(p, p2, pO2)
+	nt4.inD = true
+	// adjust the adjacent triangles for all points involved
+	p.adjacentTriangles = p.adjacentTriangles.appendT(nt1, nt2, nt3, nt4)
+	t1.inD = false
+	t2.inD = false
+	p1.adjacentTriangles = p1.adjacentTriangles.remove(t1)
+	p2.adjacentTriangles = p2.adjacentTriangles.remove(t2)
+	pO1.adjacentTriangles = pO1.adjacentTriangles.remove(t1, t2)
+	pO2.adjacentTriangles = pO2.adjacentTriangles.remove(t1, t2)
+	p1.adjacentTriangles = p1.adjacentTriangles.appendT(nt1, nt3)
+	p2.adjacentTriangles = p2.adjacentTriangles.appendT(nt2, nt4)
+	pO1.adjacentTriangles = pO1.adjacentTriangles.appendT(nt1, nt2)
+	pO2.adjacentTriangles = pO2.adjacentTriangles.appendT(nt3, nt4)
+	if d.hierarchy {
+		t1.children = append(t1.children, nt1, nt3)
+		t2.children = append(t2.children, nt2, nt4)
+	}
+	d.triangles = append(d.triangles, nt1, nt2, nt3, nt4)
+	// change the edges so it is a valid delaunay triangulation
+	d.validateEdge(nt1, p)
+	d.validateEdge(nt2, p)
+	d.validateEdge(nt3, p)
+	d.validateEdge(nt4, p)
+}
+
+// validateEdge turns triangle into a valid delaunay triangle
```

Comment:
The key operation this does is to swap the edge in t opposite to p so as to make it locally optimal or Delaunay. For me the clearest name would be `flipDelaunay` (because there is the unconditional `flip` function) or `swapDelaunay` (because I'm used to swap rather than flip terminology).